### PR TITLE
Serialization: Recover from failure to deserialize inheritance clause entries

### PR DESCRIPTION
**Pull request by slavapestov**
... by dropping them.

Fixes <rdar://problem/50473619>.

### CODE REVIEW

File: lib/Serialization/Deserialization.cpp [hunk 1]
```diff
@@ -3536,8 +3546,8 @@ class swift::DeclDeserializer {
 
     theEnum->computeType();
 
-    handleInherited(theEnum,
-                    rawInheritedAndDependencyIDs.slice(0, numInheritedTypes));
+    auto rawInheritedIDs = rawInheritedAndDependencyIDs.slice(0, numInherited);
+    handleInherited(theEnum, rawInheritedIDs);
```
Comment (jrose-apple) [inline lib/Serialization/Deserialization.cpp]: I just merged the class and struct versions for this, so you can switch to it if you want.

File: test/Serialization/Recovery/private-conformances.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+#endif
+
+#elseif CONFORMS_LIB
+import protocol_lib
+
+public protocol PublicProtocol {}
+
+public struct S : PrivateProtocol, PublicProtocol {}
+public enum E : PrivateProtocol, PublicProtocol {}
+public class C : PrivateProtocol, PublicProtocol {}
```
Comment (jrose-apple) [inline test/Serialization/Recovery/private-conformances.swift]: Needs an extension test case that doesn't involve nested types. I wonder if just asking `swift-ide-test` to print the module will have the right balance of "deserialize all the things, but don't do any lookups".

File: lib/Serialization/Deserialization.cpp [hunk 3]
```diff
@@ -564,21 +599,16 @@ ProtocolConformanceRef ModuleFile::readConformance(
...
-    // back to the current module like this.
     if (!module)
-      module = getAssociatedModule();
+      return llvm::make_error<XRefNonLoadedModuleError>(getIdentifier(moduleID));
 
     SmallVector<ProtocolConformance *, 2> conformances;
     nominal->lookupConformance(module, proto, conformances);
     PrettyStackTraceModuleFile traceMsg(
         "If you're seeing a crash here, check that your SDK and dependencies "
         "are at least as new as the versions used to build", *this);
```
Comment (jrose-apple) [inline lib/Serialization/Deserialization.cpp]: This can go away now, since it's going to recover.

File: lib/Serialization/Deserialization.cpp [hunk 4]
```diff
@@ -5439,7 +5488,12 @@ ModuleFile::loadAllConformances(const Decl *D, uint64_t contextData,
   DeclTypeCursor.JumpToBit(bitPosition);
 
   while (numConformances--) {
-    auto conf = readConformance(DeclTypeCursor);
+    auto maybeConf = readConformanceChecked(DeclTypeCursor);
+    if (!maybeConf) {
+      llvm::consumeError(maybeConf.takeError());
```
Comment (jrose-apple) [inline lib/Serialization/Deserialization.cpp]: This absolutely can't happen until we can drop BoundGenericTypes that rely on the conformance. :-(

**Review by jrose-apple — CHANGES_REQUESTED** [review]
I didn't read the third commit too closely, but we can't drop any conformances until we can detect that when deserializing BoundGenericTypes. I think we have to stick to the first commit for now.

**Comment by slavapestov** [comment]
Here's a new version with the first commit only, and a test case that use swift-ide-test to trigger extension deserialization but not member lookup.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
Can you make sure this *doesn't* recover transparently for protocols? I don't want to introduce miscompiles.

**Comment by slavapestov** [comment]
@jrose-apple Done.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Review by jrose-apple — APPROVED** [review]
Thank you for taking this! And I look forward to doing the more ambitious stuff in the future (maybe some of it on master now, even).

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Pull request merged**